pkg/foxtrot/db.go

```diff
@@ -53,18 +54,96 @@ func (db *db) setupSchema() error {
-	selectUserStmt := "SELECT password_hash FROM users WHERE name=?"
-	err := db.conn.QueryRowContext(ctx, selectUserStmt, name).Scan(&u.passwordHash)
+	stmt := "SELECT password_hash FROM users WHERE name=?"
+	err := db.conn.QueryRowContext(ctx, stmt, name).Scan(&u.passwordHash)
 	if err != nil {
 		return nil, fmt.Errorf("db: cannot get user '%s': %w", name, err)
 	}
 	return &u, nil
 }
 
 func (db *db) createUser(ctx context.Context, u *User) error {
-	insertUserStr := "INSERT INTO users(name, password_hash, avatar) VALUES( ?, ?, ? )"
-	if _, err := db.conn.ExecContext(ctx, insertUserStr, u.Name, u.passwordHash, u.avatar); err != nil {
+	stmt := "INSERT INTO users(name, password_hash, avatar) VALUES( ?, ?, ? )"
+	if _, err := db.conn.ExecContext(ctx, stmt, u.Name, u.passwordHash, u.avatar); err != nil {
 		return fmt.Errorf("db: cannot create user '%s': %w", u.Name, err)
 	}
 	return nil
 }
+
+func (db *db) getRoom(ctx context.Context, name string) (*Room, error) {
+	r := Room{Name: name}
+	stmt := "SELECT name FROM rooms WHERE name=?"
+	err := db.conn.QueryRowContext(ctx, stmt, name).Scan(&r.Name)
+	if err != nil {
+		return nil, fmt.Errorf("db: cannot get room '%s': %w", name, err)
+	}
+	return &r, nil
+}
+
+func (db *db) createRoom(ctx context.Context, r *Room) error {
+	stmt := "INSERT INTO rooms(name) VALUES( ? )"
+	if _, err := db.conn.ExecContext(ctx, stmt, r.Name); err != nil {
+		return fmt.Errorf("db: cannot create room '%s': %w", r.Name, err)
+	}
+	return nil
+}
+
+// queryMessages returns a list Messages for given room. A maximum of
+// limit messages is returned, or all messages if limit is set to -1
+// Only messages the came before given beforeID are returned or messages
+// up until the most recent one if beforeID is -1.
+func (db *db) queryMessages(ctx context.Context, room string, beforeID, limit int) ([]*Message, error) {
+	stmt := `SELECT id, content, created_at, room, author
+	         FROM messages
+			 WHERE room = ?`
+	args := []interface{}{room}
+	if beforeID != -1 {
+		stmt += " AND id < ?"
+		args = append(args, beforeID)
+	}
+	stmt += " ORDER BY id DESC"
+	if limit != -1 {
+		stmt += " LIMIT ?"
+		args = append(args, limit)
+	}
+	rows, err := db.conn.QueryContext(ctx, stmt, args...)
+	if err != nil {
+		return nil, fmt.Errorf("db: query messages for room '%s' before ID '%d': execute query: %w", room, beforeID, err)
+	}
+	defer rows.Close() //nolint:errcheck
+	messages, err := rowsToMessages(rows)
+	if err != nil {
+		return nil, fmt.Errorf("db: query messages for room '%s' before '%d': %w", room, beforeID, err)
+	}
+	return messages, nil
+}
+
+func rowsToMessages(rows *sql.Rows) ([]*Message, error) {
+	var messages []*Message
+	for rows.Next() {
+		var m Message
+		if err := rows.Scan(&m.ID, &m.Content, &m.CreatedAt, &m.Room, &m.Author); err != nil {
+			return nil, fmt.Errorf("scan row: %w", err)
+		}
+		messages = append(messages, &m)
+	}
+	// If the database is being written to ensure to check for Close
```

Comment:
remove "ensure to" from this comment. I had trouble parsing it. Was it "if the database is being written, to ensure..." or "if the database is being written to, ensure ..."

Reply:
no idea, copied it from the example ;) 
https://golang.org/pkg/database/sql/#example_DB_QueryRowContext